common/logging/src/async_record.rs

```diff
@@ -0,0 +1,309 @@
+//! An object that can be used to pass through a channel and be cloned. It can therefore be used
+//! via the broadcast channel.
+
+use parking_lot::Mutex;
+use serde::ser::SerializeMap;
+use serde::serde_if_integer128;
+use serde::Serialize;
+use slog::{BorrowedKV, Key, Level, OwnedKVList, Record, RecordStatic, Serializer, SingleKV, KV};
+use std::cell::RefCell;
+use std::fmt;
+use std::fmt::Write;
+use std::sync::Arc;
+use take_mut::take;
+
+thread_local! {
+    static TL_BUF: RefCell<String> = RefCell::new(String::with_capacity(128))
+}
+
+/// Serialized record.
+#[derive(Clone)]
+pub struct AsyncRecord {
+    msg: String,
+    level: Level,
+    location: Box<slog::RecordLocation>,
+    tag: String,
+    logger_values: OwnedKVList,
+    kv: Arc<Mutex<dyn KV + Send>>,
+}
+
+impl AsyncRecord {
+    /// Serializes a `Record` and an `OwnedKVList`.
+    pub fn from(record: &Record, logger_values: &OwnedKVList) -> Self {
+        let mut ser = ToSendSerializer::new();
+        record
+            .kv()
+            .serialize(record, &mut ser)
+            .expect("`ToSendSerializer` can't fail");
```

Comment:
I don't know enough to know that this will never panic. If you're certain it won't panic, then it's fine by me.

A non-panic alternative would be to return an `Option` here and then just don't send a message down the channel in `impl Drain for SSELoggingComponents` if this returns `None`.

My concern here would be someone figuring out how to make some graffiti/peer-id/etc that can crash our logger.

Reply:
Yeah, so slog has this `Serializer` trait. The trait returns an `slog::Result`. 

If you look here: https://github.com/AgeManning/lighthouse/blob/sse-logs/common/logging/src/async_record.rs#LL68C1-L68C1

This implements every single method of the trait. Every singe function returns Ok(()), so its impossible to return an Err. The reason it has to return a Result is just to match the trait return type. 

The slog_async crate (that we use for all our logging) has the same logic: https://docs.rs/slog-async/latest/src/slog_async/lib.rs.html#480

tldr; we're either all good, or we're in real big trouble ;P

Reply:
Fine by me!